### PR TITLE
feat: retry failed workflow run

### PR DESCRIPTION
**Pull request by mping-exo**
Start to sketch retry impl

### CODE REVIEW

File: src/ablauf/job.clj [hunk 1]
```diff
@@ -154,6 +154,14 @@
                :pred #(node/pending-and-idempotent? %1)
                :action #(dissoc %1 :exec/result)))
 
+(defn- idempotent->unstarted! [job]
+  ;; failed leafs can also be retried
+  (update-tree job
+               :pred #(and (node/idempotent? %1)
+                           (or (node/pending? %1)
+                               (node/failed? %1)))
```
Comment (pyr) [inline src/ablauf/job.clj:162]: What is the reasoning for marking a failed one as unstarted?

File: src/ablauf/job/sql.clj [hunk 2]
```diff
@@ -268,6 +269,39 @@
           wid (insert-workflow tx jobstore ast uuid owner system type metadata)]
       (insert-workflow-restart tx wid uuid []))))
 
+(defn retry
```
Comment (pyr) [inline src/ablauf/job/sql.clj:272]: In which case do you expect consumers to call `retry`?

Comment (mping-exo) [reply src/ablauf/job/sql.clj:272]: For jobs that depend on some external condition (eg calling a service that is in maintenance), and should be retried after this condition changed, mostly likely the retry will be triggered manually.

**Review by pyr — APPROVED** [review]

File: src/ablauf/job.clj (hunk 1 above)
Comment (mping-exo) [reply src/ablauf/job.clj:162]: To ensure that the leaf can be eligible for execution again.

**Comment by mping-exo** [comment]
`retry` is meant to be called when a workflow run fails due to some external conditions (eg failed network, service is down, etc).
Clients can submit a new workflow run, but that would re-execute all leafs before the failure; alternatively, they can `retry` a failed workflow run - the previous leafs would have to either succeeded or marked as idempotent (for actual retry).

**Review by tdantas — APPROVED** [review]

**Pull request merged**